From 004d50e68c891c03db12cdc967f5e86c390ada13 Mon Sep 17 00:00:00 2001
From: Philip Withnall <philip@tecnocode.co.uk>
Date: Thu, 4 Jul 2024 19:35:47 +0000
Subject: [PATCH 11/11] Merge branch 'pr/fix_some_windows_testsuite_failures'
 into 'main'

tests: fix some Windows testsuite failures

Closes #3370

See merge request GNOME/glib!4125

(cherry picked from commit 7878938e9aaef7a83ec4c58302b17ddbc080b39b)
---
 gio/tests/meson.build | 10 ++++++++--
 glib/meson.build      | 12 ++++++++----
 2 files changed, 16 insertions(+), 6 deletions(-)

diff --git a/gio/tests/meson.build b/gio/tests/meson.build
index 3bfb33307..92d8d6a28 100644
--- a/gio/tests/meson.build
+++ b/gio/tests/meson.build
@@ -835,7 +835,7 @@ if not meson.is_cross_build()
 
   test_gresource = custom_target('test.gresource',
     input : 'test.gresource.xml',
-    depends : test_generated_txt,
+    depends : [test_generated_txt, gspawn_helpers],
     output : 'test.gresource',
     command : [glib_compile_resources,
                compiler_type,
@@ -850,6 +850,7 @@ if not meson.is_cross_build()
 
   test_resources2_c = custom_target('test_resources2.c',
     input : 'test3.gresource.xml',
+    depends : [gspawn_helpers],
     output : 'test_resources2.c',
     command : [glib_compile_resources,
                compiler_type,
@@ -863,6 +864,7 @@ if not meson.is_cross_build()
 
   test_resources2_h = custom_target('test_resources2.h',
     input : 'test3.gresource.xml',
+    depends : [gspawn_helpers],
     output : 'test_resources2.h',
     command : [glib_compile_resources,
                compiler_type,
@@ -876,7 +878,7 @@ if not meson.is_cross_build()
 
   test_resources_c = custom_target('test_resources.c',
     input : 'test2.gresource.xml',
-    depends : big_test_resource,
+    depends : [big_test_resource, gspawn_helpers],
     output : 'test_resources.c',
     command : [glib_compile_resources,
                compiler_type,
@@ -890,6 +892,7 @@ if not meson.is_cross_build()
 
   digit_test_resources_c = custom_target('digit_test_resources.c',
     input : '111_digit_test.gresource.xml',
+    depends : [gspawn_helpers],
     output : 'digit_test_resources.c',
     command : [glib_compile_resources,
                compiler_type,
@@ -903,6 +906,7 @@ if not meson.is_cross_build()
 
   digit_test_resources_h = custom_target('digit_test_resources.h',
     input : '111_digit_test.gresource.xml',
+    depends : [gspawn_helpers],
     output : 'digit_test_resources.h',
     command : [glib_compile_resources,
                compiler_type,
@@ -955,6 +959,7 @@ if not meson.is_cross_build()
 
     test_gresource_binary = custom_target('test5.gresource',
       input : 'test5.gresource.xml',
+      depends : [gspawn_helpers],
       output : 'test5.gresource',
       command : [glib_compile_resources,
                  compiler_type,
@@ -970,6 +975,7 @@ if not meson.is_cross_build()
     # Create resource data file
     test_resources_binary_c = custom_target('test_resources_binary.c',
       input : 'test5.gresource.xml',
+      depends : [gspawn_helpers],
       output : 'test_resources_binary.c',
       command : [glib_compile_resources,
                  compiler_type,
diff --git a/glib/meson.build b/glib/meson.build
index d2efebadc..544f520e8 100644
--- a/glib/meson.build
+++ b/glib/meson.build
@@ -462,26 +462,30 @@ pkg.generate(libglib,
 )
 meson.override_dependency('glib-2.0', libglib_dep)
 
+# FIXME: This needs to be added to the `depends` of every invocation of
+# glib_compile_resources in the build process. We can drop this when
+# https://github.com/mesonbuild/meson/issues/13385 is fixed.
+gspawn_helpers = []
 # On Windows, glib needs a spawn helper for g_spawn* API
 if host_system == 'windows'
   if host_machine.cpu_family() == 'x86'
-    executable('gspawn-win32-helper', 'gspawn-win32-helper.c',
+    gspawn_helpers += executable('gspawn-win32-helper', 'gspawn-win32-helper.c',
       install : true,
       win_subsystem : 'windows',
       include_directories : configinc,
       dependencies : [libglib_dep])
-    executable('gspawn-win32-helper-console', 'gspawn-win32-helper.c',
+    gspawn_helpers += executable('gspawn-win32-helper-console', 'gspawn-win32-helper.c',
       install : true,
       c_args : ['-DHELPER_CONSOLE'],
       include_directories : configinc,
       dependencies : [libglib_dep])
   else
-    executable('gspawn-win64-helper', 'gspawn-win32-helper.c',
+    gspawn_helpers += executable('gspawn-win64-helper', 'gspawn-win32-helper.c',
       install : true,
       win_subsystem : 'windows',
       include_directories : configinc,
       dependencies : [libglib_dep])
-    executable('gspawn-win64-helper-console', 'gspawn-win32-helper.c',
+    gspawn_helpers += executable('gspawn-win64-helper-console', 'gspawn-win32-helper.c',
       install : true,
       c_args : ['-DHELPER_CONSOLE'],
       include_directories : configinc,
-- 
2.44.0.windows.1

